src/intrinsics.jl

```diff
@@ -0,0 +1,357 @@
+    ```
+"""
+function get_local_size end
+
+"""
+    get_local_id()::@NamedTuple{x::Int, y::Int, z::Int}
+
+Returns the unique local work-item ID.
+
+!!! note
+    1-based.
+
+!!! note
+    Backend implementations **must** implement:
+    ```
+    @device_override get_local_id()::@NamedTuple{x::Int, y::Int, z::Int}
+    ```
+"""
+function get_local_id end
+
+"""
+    get_num_groups()::@NamedTuple{x::Int, y::Int, z::Int}
+
+Returns the number of groups.
+
+!!! note
+    Backend implementations **must** implement:
+    ```
+    @device_override get_num_groups()::@NamedTuple{x::Int, y::Int, z::Int}
+    ```
+"""
+function get_num_groups end
+
+"""
+    get_group_id()::@NamedTuple{x::Int, y::Int, z::Int}
+
+Returns the unique group ID.
+
+!!! note
+    1-based.
+
+!!! note
+    Backend implementations **must** implement:
+    ```
+    @device_override get_group_id()::@NamedTuple{x::Int, y::Int, z::Int}
+    ```
+"""
+function get_group_id end
+
+"""
+    localmemory(T, dims)
+
+Declare memory that is local to a workgroup.
+
+!!! note
+    Backend implementations **must** implement:
+    ```
+    @device_override localmemory(T::DataType, ::Val{Dims}) where {T, Dims}
+    ```
+    As well as the on-device functionality.
+"""
+localmemory(::Type{T}, dims) where {T} = localmemory(T, Val(dims))
+
+"""
+    barrier()
+
+After a `barrier()` call, all read and writes to global and local memory
+from each thread in the workgroup are visible in from all other threads in the
+workgroup.
+
+!!! note
+    `barrier()` must be encountered by all workitems of a work-group executing the kernel or by none at all.
+
+!!! note
+    Backend implementations **must** implement:
+    ```
+    @device_override barrier()
+    ```
+"""
+function barrier()
+    error("Group barrier used outside kernel or not captured")
+end
+
+"""
+    _print(args...)
+
+    Overloaded by backends to enable `KernelAbstractions.@print`
+    functionality.
+
+!!! note
+    Backend implementations **must** implement:
+    ```
+    @device_override _print(args...)
+    ```
+    If the backend does not support printing,
+    define it to return `nothing`.
+"""
+@generated function _print(items...)
+    str = ""
+    args = []
+
+    for i in 1:length(items)
+        item = :(items[$i])
+        T = items[i]
+        if T <: Val
+            item = QuoteNode(T.parameters[1])
+        end
+        push!(args, item)
+    end
+
+    return quote
+        print($(args...))
+    end
+end
+
+
+"""
+    KIKernel{Backend, BKern}
+
+KIKernel closure struct that is used to represent the backend
+kernel on the host.
+
+!!! note
+    Backend implementations **must** implement:
+    ```
+    (kernel::KIKernel{<:NewBackend})(args...; numworkgroups=nothing, workgroupsize=nothing, kwargs...)
+    ```
+    As well as the on-device functionality.
+"""
+struct KIKernel{B, Kern}
+    backend::B
+    kern::Kern
+end
+
+"""
+    kernel_max_work_group_size(backend, kern; [max_work_items::Int])::Int
```

Comment:
Do we need `backend` as a first argument?

Reply:
Nope!